transaction/src/transaction.rs

```diff
@@ -79,35 +79,27 @@ impl Transaction {
     }
 
     // TODO: make sure payloads from Swap actions included
-    pub fn note_payloads(&self) -> Vec<NotePayload> {
-        self.transaction_body
-            .actions
-            .iter()
-            .filter_map(|action| {
-                if let Action::Output(output) = action {
-                    Some(output.body.note_payload.clone())
-                } else {
-                    None
-                }
-            })
-            .collect()
+    pub fn note_payloads(&self) -> impl Iterator<Item = &NotePayload> {
+        self.actions().filter_map(|action| {
+            if let Action::Output(output) = action {
+                Some(&output.body.note_payload)
+            } else {
+                None
+            }
+        })
```

Comment:
This reduces the number of intermediate allocations by returning an iterator rather than a `Vec`; the consumer can `.collect()` it if desired.